app/controllers/admin/products_v3_controller.rb

```diff
@@ -2,6 +2,101 @@
 
 module Admin
   class ProductsV3Controller < Spree::Admin::BaseController
-    def index; end
+    def index
+      fetch_products
+      render :index, locals: { producers:, categories: }
+    end
+
+    def current_user
+      spree_current_user
+    end
+
+    def producers
+      producers = OpenFoodNetwork::Permissions.new(current_user)
+        .managed_product_enterprises.is_primary_producer.by_name
+      producers.map { |p| [p.name, p.id] }
+    end
+
+    def categories
+      Spree::Taxon.order(:name).map { |c| [c.name, c.id] }
+    end
+
+    def fetch_products
```

Comment:
I guess this has been copy/pasted from reflex then?

Reply:
Yup, that's right. I think it's time I tried moving this to a ViewComponent so it can be shared between the controller and reflex. 🤔